File: Windows/Atlas.Windows/Engine/Report.cs
using System.Text.Json.Serialization;

namespace Atlas.Windows.Engine;

/// <summary>
/// The wire format atlas-engine produces, mirrored field for field.
///
/// The Swift side is the authority: Sources/AtlasCLI/Report.swift defines
/// this, and <see cref="Report.Schema"/> is checked on load so a mismatched
/// pair says so rather than misreading the numbers.
///
/// Indices are the vocabulary, exactly as in the graph. A symbol's
/// <c>File</c> indexes <c>Files</c>; a call's <c>F</c>/<c>T</c> index
/// <c>Symbols</c>; a diagram card's <c>Node</c> indexes
/// <c>Diagram.Nodes</c>. Nothing is looked up by string.
///
/// Optionals are omitted rather than written as null, which is why the
/// nullable types here matter: a missing Container, Symbol, From, Diagram or
/// Drift simply is not in the document.
/// </summary>
public sealed record Report
{
    /// <summary>The schema this client understands.</summary>
    public const int ExpectedSchema = 2;

    [JsonPropertyName("schema")]    public int Schema { get; init; }
    [JsonPropertyName("project")]   public ProjectInfo Project { get; init; } = new();
    [JsonPropertyName("files")]     public IReadOnlyList<FileEntry> Files { get; init; } = [];
    [JsonPropertyName("symbols")]   public IReadOnlyList<SymbolEntry> Symbols { get; init; } = [];
    [JsonPropertyName("calls")]     public IReadOnlyList<Call> Calls { get; init; } = [];
    [JsonPropertyName("fileEdges")] public IReadOnlyList<FileEdge> FileEdges { get; init; } = [];
    [JsonPropertyName("hubs")]      public IReadOnlyList<int> Hubs { get; init; } = [];
    [JsonPropertyName("map")]       public MapInfo? Map { get; init; }
    [JsonPropertyName("issues")]    public IReadOnlyList<IssueEntry> Issues { get; init; } = [];
    [JsonPropertyName("route")]     public IReadOnlyList<RouteStep> Route { get; init; } = [];
    [JsonPropertyName("drift")]     public DriftReport? Drift { get; init; }
}

public sealed record ProjectInfo
{
    [JsonPropertyName("name")]         public string Name { get; init; } = "";
    [JsonPropertyName("root")]         public string Root { get; init; } = "";
    /// <summary>Stable key, e.g. "web framework". Match on this, not the label.</summary>
    [JsonPropertyName("kind")]         public string Kind { get; init; } = "";
    /// <summary>The same thing in the language the report was asked for.</summary>
    [JsonPropertyName("kindLabel")]    public string KindLabel { get; init; } = "";
    /// <summary>What that kind of thing is, in prose — the opening sentence.</summary>
    [JsonPropertyName("kindExplanation")] public string KindExplanation { get; init; } = "";
    [JsonPropertyName("fileCount")]    public int FileCount { get; init; }
    [JsonPropertyName("lineCount")]    public int LineCount { get; init; }
    [JsonPropertyName("symbolCount")]  public int SymbolCount { get; init; }
    [JsonPropertyName("callCount")]    public int CallCount { get; init; }
    [JsonPropertyName("parseSeconds")] public double ParseSeconds { get; init; }
    [JsonPropertyName("analysedAt")]   public DateTimeOffset AnalysedAt { get; init; }
    [JsonPropertyName("languages")]    public IReadOnlyList<LanguageCount> Languages { get; init; } = [];
}

public sealed record LanguageCount
{
    [JsonPropertyName("id")]    public string Id { get; init; } = "";
    [JsonPropertyName("name")]  public string Name { get; init; } = "";
    [JsonPropertyName("files")] public int Files { get; init; }
}

public sealed record FileEntry
{
    /// <summary>Relative to the project root, always separated by '/'.</summary>
    [JsonPropertyName("path")]      public string Path { get; init; } = "";
    [JsonPropertyName("name")]      public string Name { get; init; } = "";
    [JsonPropertyName("directory")] public string Directory { get; init; } = "";
    [JsonPropertyName("language")]  public string Language { get; init; } = "";
}

public sealed record SymbolEntry
{
    [JsonPropertyName("name")]       public string Name { get; init; } = "";
    [JsonPropertyName("container")]  public string? Container { get; init; }
    [JsonPropertyName("display")]    public string Display { get; init; } = "";
    [JsonPropertyName("kind")]       public string Kind { get; init; } = "";
    [JsonPropertyName("language")]   public string Language { get; init; } = "";
    /// <summary>Index into <see cref="Report.Files"/>; -1 when it has no file.</summary>
    [JsonPropertyName("file")]       public int File { get; init; }
    [JsonPropertyName("line")]       public int Line { get; init; }
    [JsonPropertyName("endLine")]    public int EndLine { get; init; }
    [JsonPropertyName("fanIn")]      public int FanIn { get; init; }
    [JsonPropertyName("fanOut")]     public int FanOut { get; init; }
    [JsonPropertyName("external")]   public bool External { get; init; }
    [JsonPropertyName("branches")]   public int Branches { get; init; }
    [JsonPropertyName("nesting")]    public int Nesting { get; init; }
    /// <summary>easy | moderate | hard</summary>
    [JsonPropertyName("difficulty")] public string Difficulty { get; init; } = "";

    public int Span => Math.Max(1, EndLine - Line + 1);
}

/// <summary>
/// One call edge. Short names because a large project has tens of thousands
/// of these and they dominate the document.
/// </summary>
public sealed record Call
{
    [JsonPropertyName("f")] public int From { get; init; }
    [JsonPropertyName("t")] public int To { get; init; }
    [JsonPropertyName("n")] public int Sites { get; init; }
}

/// <summary>A file-to-file dependency taken from imports rather than calls.</summary>
public sealed record FileEdge
{
    [JsonPropertyName("f")] public int From { get; init; }
    [JsonPropertyName("t")] public int To { get; init; }
}

/// <summary>
/// The call ladder, already placed by the engine.
///
/// The layout travels with the data because it is deliberate — depth by
/// longest path, then barycentre sweeps to reduce crossings — and because the
/// picture is the product. Two clients each deriving their own would draw the
/// same repository two different ways. Coordinates are absolute on a canvas of
/// <see cref="Canvas"/> size, so this client only pans and zooms.
/// </summary>
public sealed record MapInfo
{
    [JsonPropertyName("canvas")]  public CanvasSize Canvas { get; init; } = new();
    [JsonPropertyName("nodes")]   public IReadOnlyList<MapNode> Nodes { get; init; } = [];
    [JsonPropertyName("boxes")]   public IReadOnlyList<Box> Boxes { get; init; } = [];
    /// <summary>Node indices per column, in the order they are stacked.</summary>
    [JsonPropertyName("columns")] public IReadOnlyList<IReadOnlyList<int>> Columns { get; init; } = [];
    [JsonPropertyName("edges")]   public IReadOnlyList<MapEdge> Edges { get; init; } = [];
    /// <summary>Indices into <see cref="Nodes"/>.</summary>
    [JsonPropertyName("cycles")]  public IReadOnlyList<IReadOnlyList<int>> Cycles { get; init; } = [];
    /// <summary>The same dependency data arranged as a grid rather than a ladder.</summary>
    [JsonPropertyName("matrix")]  public MatrixInfo Matrix { get; init; } = new();
}

/// <summary>Row and column order for the dependency matrix: rows call columns.</summary>
public sealed record MatrixInfo
{
    /// <summary>Node indices, in the order they are laid out.</summary>
    [JsonPropertyName("order")]     public IReadOnlyList<int> Order { get; init; } = [];
    [JsonPropertyName("districts")] public IReadOnlyList<MatrixDistrict> Districts { get; init; } = [];
}

public sealed record MatrixDistrict
{
    /// <summary>interface | logic | data — a key, printed by the client.</summary>
    [JsonPropertyName("key")]   public string Key { get; init; } = "";
    /// <summary>Where this district starts in <see cref="MatrixInfo.Order"/>.</summary>
    [JsonPropertyName("start")] public int Start { get; init; }
}

public sealed record CanvasSize
{
    [JsonPropertyName("w")] public double Width { get; init; }
    [JsonPropertyName("h")] public double Height { get; init; }
}

public sealed record MapNode
{
    /// <summary>Index into <see cref="Report.Files"/>.</summary>
    [JsonPropertyName("file")]        public int File { get; init; }
    [JsonPropertyName("path")]        public string Path { get; init; } = "";
    [JsonPropertyName("name")]        public string Name { get; init; } = "";
    /// <summary>Stable key — drives the district rule down the box's left edge.</summary>
    [JsonPropertyName("layer")]       public string Layer { get; init; } = "";
    [JsonPropertyName("layerLabel")]  public string LayerLabel { get; init; } = "";
    [JsonPropertyName("language")]    public string Language { get; init; } = "";
    /// <summary>Indices into <see cref="Report.Symbols"/>, most connected first.</summary>
    [JsonPropertyName("symbols")]     public IReadOnlyList<int> Symbols { get; init; } = [];
    [JsonPropertyName("symbolCount")] public int SymbolCount { get; init; }
    [JsonPropertyName("lines")]       public int Lines { get; init; }
    [JsonPropertyName("fanIn")]       public int FanIn { get; init; }
    [JsonPropertyName("fanOut")]      public int FanOut { get; init; }
}

public sealed record Box
{
    /// <summary>Index into <see cref="MapInfo.Nodes"/>.</summary>
    [JsonPropertyName("node")]   public int Node { get; init; }
    [JsonPropertyName("x")]      public double X { get; init; }
    [JsonPropertyName("y")]      public double Y { get; init; }
    [JsonPropertyName("w")]      public double Width { get; init; }
    [JsonPropertyName("h")]      public double Height { get; init; }
    [JsonPropertyName("column")] public int Column { get; init; }
    [JsonPropertyName("row")]    public int Row { get; init; }
}

/// <summary>A file-to-file dependency, weighted by how many call sites back it.</summary>
public sealed record MapEdge
{
    [JsonPropertyName("f")]      public int From { get; init; }
    [JsonPropertyName("t")]      public int To { get; init; }
    [JsonPropertyName("weight")] public int Weight { get; init; }
}

public sealed record IssueEntry
{
    [JsonPropertyName("id")]       public int Id { get; init; }
    [JsonPropertyName("kind")]     public string Kind { get; init; } = "";
    /// <summary>low | medium | high</summary>
    [JsonPropertyName("severity")] public string Severity { get; init; } = "";
    [JsonPropertyName("title")]    public string Title { get; init; } = "";
    [JsonPropertyName("subject")]  public string Subject { get; init; } = "";
    [JsonPropertyName("detail")]   public string Detail { get; init; } = "";
    [JsonPropertyName("advice")]   public string Advice { get; init; } = "";
    [JsonPropertyName("file")]     public string File { get; init; } = "";
    [JsonPropertyName("line")]     public int Line { get; init; }
    /// <summary>Index into <see cref="Report.Symbols"/>, when there is a jump target.</summary>
    [JsonPropertyName("symbol")]   public int? Symbol { get; init; }
}

public sealed record RouteStep
{
    /// <summary>Index into <see cref="Report.Symbols"/>.</summary>
    [JsonPropertyName("symbol")] public int Symbol { get; init; }
    /// <summary>The step that calls this one; absent for the first.</summary>
    [JsonPropertyName("from")]   public int? From { get; init; }
    /// <summary>How many symbols this step can reach, counted by the engine.</summary>
    [JsonPropertyName("reach")]  public int Reach { get; init; }
}

public sealed record DriftReport
{
    [JsonPropertyName("previousScan")] public DateTimeOffset? PreviousScan { get; init; }
    [JsonPropertyName("entries")]      public IReadOnlyList<DriftEntry> Entries { get; init; } = [];
}

public sealed record DriftEntry
{
    [JsonPropertyName("kind")]        public string Kind { get; init; } = "";
    /// <summary>Empty for the aggregate kinds, where the count is the whole story.</summary>
    [JsonPropertyName("subject")]     public string Subject { get; init; } = "";
    [JsonPropertyName("delta")]       public int Delta { get; init; }
    [JsonPropertyName("detail")]      public string Detail { get; init; } = "";
    /// <summary>The change written out as a sentence, composed by the engine.</summary>
    [JsonPropertyName("note")]        public string Note { get; init; } = "";
    [JsonPropertyName("regression")]  public bool Regression { get; init; }
    [JsonPropertyName("improvement")] public bool Improvement { get; init; }
}

File: Sources/AtlasCLI/Report.swift
import Foundation

/// The wire format between the engine and a user interface.
///
/// Written out as explicit DTOs rather than by making the graph types
/// `Codable`. Two reasons: the analysis types stay free to change shape
/// without silently breaking a shipped UI, and the format can say things the
/// in-memory model leaves implicit — a localised issue title, a difficulty
/// band, a colour-bearing layer — so the Windows client does not have to
/// re-derive them.
///
/// Indices are the vocabulary here, exactly as they are in the graph. A
/// symbol's `file` indexes `files`; a call's `f`/`t` index `symbols`; a
/// diagram card's `node` indexes `diagram.nodes`. Nothing is looked up by
/// string.
///
/// Optionals are **omitted rather than written as `null`** — that is
/// `JSONEncoder`'s default, and it keeps a large report meaningfully
/// smaller. A missing `container`, `symbol`, `from`, `diagram` or `drift`
/// means absent, and a client should read them as nullable.
struct Report: Encodable {

    /// Bumped when a field changes meaning or disappears, so a mismatched
    /// client can say so instead of misreading the numbers.
    static let schema = 2

    var schema: Int = Report.schema
    var project: Project
    var files: [FileEntry]
    var symbols: [SymbolEntry]
    var calls: [Call]
    var fileEdges: [FileEdge]
    var hubs: [Int]
    var map: Map?
    var issues: [IssueEntry]
    var route: [RouteStep]
    var drift: DriftReport?

    // MARK: - Project

    struct Project: Encodable {
        var name: String
        var root: String
        var kind: String            // ProjectKind.rawValue — stable key
        var kindLabel: String       // the same, in the requested language
        var kindExplanation: String // what that kind of thing *is*, in prose
        var fileCount: Int
        var lineCount: Int
        var symbolCount: Int
        var callCount: Int
        var parseSeconds: Double
        var analysedAt: Date
        var languages: [LanguageCount]
    }

    struct LanguageCount: Encodable {
        var id: String              // Language.rawValue
        var name: String            // Language.displayName
        var files: Int
    }

    // MARK: - Files and symbols

    struct FileEntry: Encodable {
        var path: String            // relative to root, always `/`-separated
        var name: String
        var directory: String
        var language: String
    }

    struct SymbolEntry: Encodable {
        var name: String
        var container: String?
        var display: String
        var kind: String            // SymbolKind.rawValue
        var language: String
        var file: Int               // index into `files`
        var line: Int
        var endLine: Int
        var fanIn: Int
        var fanOut: Int
        var external: Bool
        var branches: Int
        var nesting: Int
        var difficulty: String      // easy | moderate | hard
    }

    /// One call edge. Short keys because a large project has tens of
    /// thousands of these and they dominate the file size.
    struct Call: Encodable {
        var f: Int                  // caller, index into `symbols`
        var t: Int                  // callee
        var n: Int                  // distinct call sites
    }

    /// A file-to-file dependency taken from imports rather than calls.
    struct FileEdge: Encodable {
        var f: Int                  // index into `files`
        var t: Int
    }

    // MARK: - Map

    /// The call ladder, already placed.
    ///
    /// The layout travels with the data because it is deliberate — depth by
    /// longest path, then barycentre sweeps to reduce crossings — and because
    /// the picture is the product. Two clients each deriving their own would
    /// draw the same repository two different ways. Coordinates are absolute
    /// on a canvas of `canvas` size, so a client only pans and zooms.
    struct Map: Encodable {
        var canvas: Size
        var nodes: [Node]
        var boxes: [Box]
        /// Node indices per column, in the order they are stacked.
        var columns: [[Int]]
        var edges: [Edge]
        var cycles: [[Int]]         // indices into `nodes`

        struct Size: Encodable { var w: Double; var h: Double }

        struct Node: Encodable {
            var file: Int           // index into `files`
            var path: String
            var name: String
            var layer: String       // Layer.rawValue — drives the district rule
            var layerLabel: String
            var language: String
            var symbols: [Int]      // indices into `symbols`, most connected first
            var symbolCount: Int
            var lines: Int
            var fanIn: Int
            var fanOut: Int
        }

        struct Box: Encodable {
            var node: Int           // index into `map.nodes`
            var x: Double
            var y: Double
            var w: Double
            var h: Double
            var column: Int
            var row: Int
        }

        /// A file-to-file dependency, weighted by how many call sites back it.
        struct Edge: Encodable {
            var f: Int              // index into `map.nodes`
            var t: Int
            var weight: Int
        }
    }

    // MARK: - Findings

    struct IssueEntry: Encodable {
        var id: Int
        var kind: String            // Issue.Kind.rawValue
        var severity: String        // low | medium | high
        var title: String
        var subject: String
        var detail: String
        var advice: String
        var file: String
        var line: Int
        var symbol: Int?            // index into `symbols`, when there is one
    }

    struct RouteStep: Encodable {
        var symbol: Int             // index into `symbols`
        var from: Int?              // the step that calls it
        /// How many symbols this step can reach. Computed here because it is
        /// a graph walk, and a client that repeated it would be reimplementing
        /// the analysis it was given.
        var reach: Int
    }

    struct DriftReport: Encodable {
        var previousScan: Date?
        var entries: [Entry]

        struct Entry: Encodable {
            var kind: String        // Drift.Kind.rawValue
            var subject: String
            var delta: Int
            var detail: String
            /// The change written out as a sentence. Composed here because the
            /// wording depends on the kind *and* the numbers, which no simple
            /// template survives — and because an entry can have no subject at
            /// all, where the sentence is the whole row.
            var note: String
            var regression: Bool
            var improvement: Bool
        }
    }
}

// MARK: - Building

extension Report {

    /// Assembles the report from a finished analysis.
    static func build(graph: CodeGraph,
                      fileGraph: FileGraph,
                      layout: LadderLayout?,
                      issues: [Issue],
                      route: Route,
                      kind: ProjectKind,
                      drift: Drift?,
                      language: AppLanguage) -> Report {

        let languages = graph.languageCounts
            .sorted { $0.value > $1.value }
            .map { LanguageCount(id: $0.key.rawValue,
                                 name: $0.key.displayName,
                                 files: $0.value) }

        let project = Project(
            name: graph.projectName,
            root: graph.rootPath,
            kind: kind.rawValue,
            kindLabel: language == .uz ? kind.uz : kind.en,
            kindExplanation: kind.explanation(language: language),
            fileCount: graph.files.count,
            lineCount: graph.totalLines,
            symbolCount: graph.nodes.count,
            callCount: graph.edges.count,
            parseSeconds: graph.parseSeconds,
            analysedAt: Date(),
            languages: languages)

        let files = graph.files.map {
            FileEntry(path: $0,
                      name: P.lastComponent($0),
                      directory: P.deletingLastComponent($0),
                      language: Language.detect(path: $0)?.rawValue ?? "")
        }

        let symbols = graph.nodes.map { node in
            SymbolEntry(name: node.name,
                        container: node.container,
                        display: node.displayName,
                        kind: node.kind.rawValue,
                        language: node.language.rawValue,
                        file: node.fileIndex,
                        line: node.line,
                        endLine: node.endLine,
                        fanIn: node.fanIn,
                        fanOut: node.fanOut,
                        external: node.isExternal,
                        branches: node.branches,
                        nesting: node.maxNesting,
                        difficulty: label(node.difficulty))
        }

        let calls = graph.edges.map { Call(f: $0.from, t: $0.to, n: $0.count) }
        let fileEdges = graph.fileReferences.map { FileEdge(f: $0.from, t: $0.to) }

        let issueEntries = issues.map {
            IssueEntry(id: $0.id,
                       kind: $0.kind.rawValue,
                       severity: label($0.severity),
                       title: $0.title(language),
                       subject: $0.subject,
                       detail: $0.detail,
                       advice: $0.advice(language),
                       file: $0.file,
                       line: $0.line,
                       symbol: $0.symbolID)
        }

        let steps = route.steps.map {
            RouteStep(symbol: $0.nodeID, from: $0.reachedFrom,
                      reach: graph.reach(of: $0.nodeID))
        }

        let words = L10n(language: language)
        let driftReport = drift.map { d in
            DriftReport(previousScan: d.previousScan,
                        entries: d.entries.map {
                            DriftReport.Entry(kind: $0.kind.rawValue,
                                              subject: $0.subject,
                                              delta: $0.delta,
                                              detail: $0.detail,
                                              note: words.driftNote($0),
                                              regression: $0.isRegression,
                                              improvement: $0.isImprovement)
                        })
        }

        return Report(project: project,
                      files: files,
                      symbols: symbols,
                      calls: calls,
                      fileEdges: fileEdges,
                      hubs: graph.hubs(limit: 12),
                      map: layout.map { map(fileGraph: fileGraph, layout: $0,
                                            language: language) },
                      issues: issueEntries,
                      route: steps,
                      drift: driftReport)
    }

    private static func map(fileGraph: FileGraph,
                            layout: LadderLayout,
                            language: AppLanguage) -> Map {
        let nodes = fileGraph.nodes.map { node in
            Map.Node(file: node.id,
                     path: node.path,
                     name: node.name,
                     layer: node.layer.rawValue,
                     layerLabel: language == .uz ? node.layer.uz : node.layer.en,
                     language: node.language.rawValue,
                     symbols: node.symbols,
                     symbolCount: node.symbolCount,
                     lines: node.lines,
                     fanIn: node.fanIn,
                     fanOut: node.fanOut)
        }

        // Sorted by node index: `frame` is a dictionary, and its order is not
        // the same twice in a row.
        let boxes = layout.frame.keys.sorted().compactMap { index -> Map.Box? in
            guard let rect = layout.frame[index] else { return nil }
            let column = layout.columns.firstIndex { $0.contains(index) } ?? 0
            let row = layout.columns[column].firstIndex(of: index) ?? 0
            return Map.Box(node: index,
                           x: Double(rect.origin.x), y: Double(rect.origin.y),
                           w: Double(rect.size.width), h: Double(rect.size.height),
                           column: column, row: row)
        }

        return Map(canvas: Map.Size(w: Double(layout.canvas.width),
                                    h: Double(layout.canvas.height)),
                   nodes: nodes,
                   boxes: boxes,
                   columns: layout.columns,
                   edges: fileGraph.edges.map {
                       Map.Edge(f: $0.from, t: $0.to, weight: $0.weight)
                   },
                   cycles: fileGraph.cycles)
    }

    // Both enums are ordered rather than named in the model, because ordering
    // is what the app sorts by. The wire format spells them out.
    private static func label(_ d: GraphNode.Difficulty) -> String {
        switch d {
        case .easy: return "easy"
        case .moderate: return "moderate"
        case .hard: return "hard"
        }
    }

    private static func label(_ s: Issue.Severity) -> String {
        switch s {
        case .low: return "low"
        case .medium: return "medium"
        case .high: return "high"
        }
    }
}
